File: typescript/src/infra/redis-infra.ts
// Redis-backed ports: the membership cache (its coherence with Postgres is the
// G-CACHE property), per-channel unread counters, and heartbeats (writes the
// same presence key the gRPC service reads). These are the "correct" wiring the
// module assembles; their Redis client is injected.

import type Redis from 'ioredis';

import type { Heartbeats, MembershipCache, UnreadCounters } from '../seams/seams.js';

export const PRESENCE_KEY_PREFIX = 'presence:';
const MEMBERSHIP_TTL_SECONDS = 300;
const PRESENCE_TTL_SECONDS = 60;

const membersKey = (channelId: string): string => `members:${channelId}`;
const unreadKey = (userId: string, channelId: string): string => `unread:${userId}:${channelId}`;

export class RedisMembershipCache implements MembershipCache {
  constructor(private readonly client: Redis) {}

  async isMember(channelId: string, userId: string): Promise<{ cached: boolean; member: boolean }> {
    const key = membersKey(channelId);
    const exists = await this.client.exists(key);
    if (exists === 0) {
      return { cached: false, member: false };
    }
    const member = await this.client.sismember(key, userId);
    return { cached: true, member: member === 1 };
  }

  async remember(channelId: string, memberIds: string[]): Promise<void> {
    const key = membersKey(channelId);
    const pipe = this.client.multi();
    pipe.del(key);
    if (memberIds.length > 0) {
      pipe.sadd(key, ...memberIds);
      pipe.expire(key, MEMBERSHIP_TTL_SECONDS);
    }
    await pipe.exec();
  }

  async invalidate(channelId: string): Promise<void> {
    await this.client.del(membersKey(channelId));
  }
}

export class RedisUnreadCounters implements UnreadCounters {
  constructor(private readonly client: Redis) {}

  async increment(userId: string, channelId: string): Promise<void> {
    await this.client.incr(unreadKey(userId, channelId));
  }

  async reset(userId: string, channelId: string): Promise<void> {
    await this.client.del(unreadKey(userId, channelId));
  }

  async forUser(userId: string): Promise<Record<string, number>> {
    const prefix = `unread:${userId}:`;
    const out: Record<string, number> = {};
    let cursor = '0';
    do {
      const [next, keys] = await this.client.scan(cursor, 'MATCH', `${prefix}*`, 'COUNT', 100);
      cursor = next;
      for (const key of keys) {
        const channelId = key.slice(prefix.length);
        const raw = await this.client.get(key);
        out[channelId] = raw === null ? 0 : Number.parseInt(raw, 10);
      }
    } while (cursor !== '0');
    return out;
  }
}

export class RedisHeartbeats implements Heartbeats {
  constructor(private readonly client: Redis) {}

  async mark(userId: string): Promise<void> {
    await this.client.set(`${PRESENCE_KEY_PREFIX}${userId}`, '1', 'EX', PRESENCE_TTL_SECONDS);
  }
}

File: python/src/relay/infra/redis_infra.py
"""Redis-backed ports: membership cache, unread counters, heartbeats."""

from __future__ import annotations

import redis

from relay.presence import KEY_PREFIX

MEMBERSHIP_TTL = 300  # seconds
PRESENCE_TTL = 60  # seconds


def _members_key(channel_id: str) -> str:
    return "members:" + channel_id


class MembershipCache:
    """The Redis membership fast-path (members:{channelId}, a set) coupled to
    invalidation on membership writes. Its coherence with Postgres is the
    G-CACHE property."""

    def __init__(self, client: redis.Redis) -> None:
        self._redis = client

    def is_member(self, channel_id: str, user_id: str) -> tuple[bool, bool]:
        key = _members_key(channel_id)
        if self._redis.exists(key) == 0:
            return (False, False)
        return (True, bool(self._redis.sismember(key, user_id)))

    def remember(self, channel_id: str, member_ids: list[str]) -> None:
        key = _members_key(channel_id)
        pipe = self._redis.pipeline(transaction=True)
        pipe.delete(key)
        if member_ids:
            pipe.sadd(key, *member_ids)
            pipe.expire(key, MEMBERSHIP_TTL)
        pipe.execute()

    def invalidate(self, channel_id: str) -> None:
        self._redis.delete(_members_key(channel_id))


def _unread_key(user_id: str, channel_id: str) -> str:
    return f"unread:{user_id}:{channel_id}"


class UnreadCounters:
    """The Redis per-channel unread counter (unread:{userId}:{channelId})."""

    def __init__(self, client: redis.Redis) -> None:
        self._redis = client

    def increment(self, user_id: str, channel_id: str) -> None:
        self._redis.incr(_unread_key(user_id, channel_id))

    def reset(self, user_id: str, channel_id: str) -> None:
        self._redis.delete(_unread_key(user_id, channel_id))

    def for_user(self, user_id: str) -> dict[str, int]:
        prefix = f"unread:{user_id}:"
        out: dict[str, int] = {}
        for key in self._redis.scan_iter(match=prefix + "*", count=100):
            channel_id = key[len(prefix):]
            raw = self._redis.get(key)
            out[channel_id] = int(raw) if raw is not None else 0
        return out


class Heartbeats:
    """Writes the presence key the gRPC service reads (presence:{userId}, 60 s TTL)."""

    def __init__(self, client: redis.Redis) -> None:
        self._redis = client

    def mark(self, user_id: str) -> None:
        self._redis.set(KEY_PREFIX + user_id, "1", ex=PRESENCE_TTL)
